core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnReverse.java

```diff
@@ -0,0 +1,94 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND FREEDOM FROM
+ * INFRINGEMENT, AND ANY WARRANTY THAT THE DOCUMENTATION WILL CONFORM TO THE
+ * SOFTWARE, OR ANY WARRANTY THAT THE SOFTWARE WILL BE ERROR FREE.  IN NO EVENT
+ * SHALL NIST BE LIABLE FOR ANY DAMAGES, INCLUDING, BUT NOT LIMITED TO, DIRECT,
+ * INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES, ARISING OUT OF, RESULTING FROM,
+ * OR IN ANY WAY CONNECTED WITH THIS SOFTWARE, WHETHER OR NOT BASED UPON WARRANTY,
+ * CONTRACT, TORT, OR OTHERWISE, WHETHER OR NOT INJURY WAS SUSTAINED BY PERSONS OR
+ * PROPERTY OR OTHERWISE, AND WHETHER OR NOT LOSS WAS SUSTAINED FROM, OR AROSE OUT
+ * OF THE RESULTS OF, OR USE OF, THE SOFTWARE OR SERVICES PROVIDED HEREUNDER.
+ */
+
+package gov.nist.secauto.metaschema.core.metapath.function.library;
+
+import gov.nist.secauto.metaschema.core.metapath.DynamicContext;
+import gov.nist.secauto.metaschema.core.metapath.ISequence;
+import gov.nist.secauto.metaschema.core.metapath.MetapathConstants;
+import gov.nist.secauto.metaschema.core.metapath.function.FunctionUtils;
+import gov.nist.secauto.metaschema.core.metapath.function.IArgument;
+import gov.nist.secauto.metaschema.core.metapath.function.IFunction;
+import gov.nist.secauto.metaschema.core.metapath.item.IItem;
+import gov.nist.secauto.metaschema.core.util.ObjectUtils;
+
+import java.util.Collections;
+import java.util.List;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Implements <a href=
+ * "https://www.w3.org/TR/xpath-functions-31/#func-exists">fn:exists</a>.
+ */
+public final class FnReverse {
+  @NonNull
+  static final IFunction SIGNATURE_ONE_ARG = IFunction.builder()
+      .name("reverse")
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("target")
+          .type(IItem.class)
+          .zeroOrMore()
+          .build())
+      .returnType(IItem.class)
+      .returnZeroOrMore()
+      .functionHandler(FnReverse::execute)
+      .build();
+
+  private FnReverse() {
+    // disable construction
+  }
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<?> execute(@NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+    ISequence<?> target = FunctionUtils.asType(ObjectUtils.requireNonNull(arguments.get(0)));
+    return ISequence.of(fnReverse(target));
+  }
+
+  /**
+   * Identify if there is at least one item in the {@code sequence}.
+   *
+   * @param sequence
+   *          the sequence to check
+   * @return {@code true} if the sequence contains at least one item, or
+   *         {@code false} otherwise
+   */
+  public static List<? extends IItem> fnReverse(List<? extends IItem> target) {
+    if (target.size() <= 1) {
+      return target;
+    }
```

Comment:
This logic is incomplete relative to the specification.



> The function returns a sequence consisting of all items of $target whose index is less than $position, followed by all items of $target whose index is greater than $position.

> If $position is less than 1 or greater than the number of items in $target, $target is returned.

> If $target is the empty sequence, the empty sequence is returned.

Reply:
I think you meant this for `fn:remove()`, but I got the idea about needing to make a copy to mutate and not using the other. I had to double check to understand if this was wrong, but I would like to make sure I understood during final review.

Reply:
Hi @david-waltermire-nist, I am going to ask you to re-review. In the interim as we touch this up for final review, I double-checked the code is in face mutating a copy, not the original passed in as an argument (whether we were talking about `fn:remove` or the other two, we did that for `fn:insert-before` and `fn:reverse` as well.

I am going to close this comment out for now.